gadget-container/gadgets/collector-process/c/ebpf.c

```diff
@@ -0,0 +1,112 @@
+#include <linux/bpf.h>
+
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_endian.h>
+
+struct bpf_iter_meta {
+	__bpf_md_ptr(struct seq_file *, seq);
+	__u64 session_id;
+	__u64 seq_num;
+};
+
+struct bpf_iter__task {
+	struct bpf_iter_meta *meta;
+	struct task_struct *task;
+} __attribute__((preserve_access_index));
+
+struct ns_common {
+	unsigned int inum;
+} __attribute__((preserve_access_index));
+
+struct nsproxy {
+	struct mnt_namespace *mnt_ns;
+} __attribute__((preserve_access_index));
+
+struct task_struct {
+	int pid;
+	int tgid;
+	char comm[16];
+	struct nsproxy *nsproxy;
+} __attribute__((preserve_access_index));
+
+struct mnt_namespace {
+	struct ns_common ns;
+}__attribute__((preserve_access_index));
+
+struct container {
+	char ContainerID[256];
+	char KubernetesNamespace[256];
+	char KubernetesPod[256];
+	char KubernetesContainer[256];
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__type(key, __u64);
+	__type(value, struct container);
+	__uint(max_entries, 1024);
+} containers SEC(".maps");
+
+
+/* From: tools/lib/bpf/bpf_tracing.h */
+/*
+ * BPF_SEQ_PRINTF to wrap bpf_seq_printf to-be-printed values
+ * in a structure.
+ */
+#define BPF_SEQ_PRINTF(seq, fmt, args...)				    \
+	({								    \
+		_Pragma("GCC diagnostic push")				    \
+		_Pragma("GCC diagnostic ignored \"-Wint-conversion\"")	    \
+		static const char ___fmt[] = fmt;			    \
+		unsigned long long ___param[] = { args };		    \
+		_Pragma("GCC diagnostic pop")				    \
+		int ___ret = bpf_seq_printf(seq, ___fmt, sizeof(___fmt),    \
+					    ___param, sizeof(___param));    \
+		___ret;							    \
+	})
```

Comment:
https://github.com/libbpf/libbpf/commit/343f63e2453269965252f5503c86e83916a12aff is not yet available in released versions of libbpf. But I should add a comment.

Reply:
Actually, it's available in the libbpf versions we are using but in a different header. So I can delete my define and use the one in libbpf.

```diff
-#include <bpf/bpf_endian.h>
+
+/* libbpf v0.4.0 introduced BPF_SEQ_PRINTF in bpf_tracing.h.
+ * In future versions, it will be in bpf_helpers.h.
+ */
+#include <bpf/bpf_tracing.h>
```

Reply:
I see. Btw, we're using cilium/ebpf that **doesn't** depend on libbpf but we need libbpf for the bpf headers, right? I wondering if this is the common approach on this case.

Reply:
I think that's correct. cilium/ebpf does not compile anything like bcc does: it only takes the ELF objects. So it should not care that we use the headers from libbpf.

I see that the cilium/ebpf repository has some headers at examples/headers/bpf_helpers.h but they don't seem to contain what we need.

Reply:
Got it. Perhaps having a `libbpf-headers` package could help... Anyway, requiring to have libbpf installed to compile is not a big deal.